Mathlib/RingTheory/LocalProperties/Projective.lean

```diff
@@ -0,0 +1,148 @@
+      apply IsLocalizedModule.ringHom_ext I.primeCompl (LocalizedModule.mkLinearMap I.primeCompl N)
+      · exact IsLocalizedModule.map_units (LocalizedModule.mkLinearMap I.primeCompl N)
+      ext
+      simp only [LocalizedModule.map_mk, LinearMap.coe_comp, LinearMap.coe_restrictScalars,
+        Function.comp_apply, LocalizedModule.mkLinearMap_apply, LinearMap.llcomp_apply,
+        LocalizedModule.map_mk]
+    · rintro ⟨g, rfl⟩
+      obtain ⟨⟨g, s⟩, rfl⟩ :=
+        IsLocalizedModule.mk'_surjective I.primeCompl (LocalizedModule.map I.primeCompl) g
+      simp only [Function.uncurry_apply_pair, Submodule.restrictScalars_mem]
+      refine ⟨f.comp g, ⟨g, rfl⟩, s, ?_⟩
+      apply ((Module.End_isUnit_iff _).mp <| IsLocalizedModule.map_units
+         (LocalizedModule.map I.primeCompl) s).injective
+      simp only [Module.algebraMap_end_apply, ← Submonoid.smul_def, IsLocalizedModule.mk'_cancel',
+        ← LinearMap.map_smul_of_tower]
+      apply LinearMap.restrictScalars_injective R
+      apply IsLocalizedModule.ringHom_ext I.primeCompl (LocalizedModule.mkLinearMap I.primeCompl N)
+      · exact IsLocalizedModule.map_units (LocalizedModule.mkLinearMap I.primeCompl N)
+      ext
+      simp only [coe_comp, coe_restrictScalars, Function.comp_apply,
+        LocalizedModule.mkLinearMap_apply, LocalizedModule.map_mk, llcomp_apply]
+
+theorem Module.projective_of_localization_maximal (H : ∀ (I : Ideal R) (_ : I.IsMaximal),
+    Module.Projective (Localization.AtPrime I) (LocalizedModule I.primeCompl M))
+    [Module.FinitePresentation R M] : Module.Projective R M := by
+  have : Module.Finite R M := by infer_instance
+  have : (⊤ : Submodule R M).FG := this.out
+  have : ∃ (s : Finset M), _ := this
+  obtain ⟨s, hs⟩ := this
+  let N := s →₀ R
+  let f : N →ₗ[R] M := Finsupp.linearCombination R (Subtype.val : s → M)
+  have hf : Function.Surjective f := by
+    rw [← LinearMap.range_eq_top, Finsupp.range_linearCombination, Subtype.range_val]
+    convert hs
+  have (I : Ideal R) (hI : I.IsMaximal) :=
+    letI := H I hI
+    Module.projective_lifting_property (LocalizedModule.map I.primeCompl f) LinearMap.id
+    (LocalizedModule.map_surjective _ _ hf)
+  obtain ⟨g, hg⟩ := LinearMap.split_surjective_of_localization_maximal _ this
+  exact Module.Projective.of_split _ _ hg
+
+variable
+  (Rₚ : ∀ (P : Ideal R) [P.IsMaximal], Type*)
+  [∀ (P : Ideal R) [P.IsMaximal], CommRing (Rₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Algebra R (Rₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], IsLocalization.AtPrime (Rₚ P) P]
+  (Mₚ : ∀ (P : Ideal R) [P.IsMaximal], Type*)
+  [∀ (P : Ideal R) [P.IsMaximal], AddCommGroup (Mₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Module R (Mₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], Module (Rₚ P) (Mₚ P)]
+  [∀ (P : Ideal R) [P.IsMaximal], IsScalarTower R (Rₚ P) (Mₚ P)]
+  (f : ∀ (P : Ideal R) [P.IsMaximal], M →ₗ[R] Mₚ P)
+  [inst : ∀ (P : Ideal R) [P.IsMaximal], IsLocalizedModule P.primeCompl (f P)]
```

Comment:
At some point we might want to have some syntax for variable blocks like this...